### ISSUE TITLE
Slow creation of bup structs and other thoughts

### ISSUE THREAD

**Issue by aidanhs** [aidanhs/rsroll#7]
https://github.com/dpc/rdedup/commit/b1298e60e973683c6cc5aa3d24612114ac58660d seems to show a measurable performance improvement if you skip creating a new struct.

1. this change does actually change splitting behaviour if you haven't set a minimum chunk size greater than the window size of the rollsum (since your content defined chunking now depends on the content of the *previous* chunk until you've rolled over the whole window)
2. why is it so slow? It should just be a memcpy of 128 bytes
3. maybe we should expose a reset method if rust can't make it fast
4. maybe we should make an iterator with arguments for min size etc to handle the somewhat clever logic mentioned in https://github.com/dpc/rdedup/issues/88#issuecomment-317433109

cc @dpc

**Comment by dpc** [comment]
So, I think I might have confused myself. The "before" in b1298e6 is before other optimizations that I submitted to rsroll, and after is with these optimizations (but still `bup`, not `gear`) and not creating new `Bup` structure every time.

It might be just "over-optimizing". Though, it seems to me that in practice, little chunking differences in splitting behavior don't matter, especially when using big chunk sizes like 1MB.

However, considering that `gear` seem to be much faster, and has much smaller state to reset, I might stop using `bup` at all.

**Comment by dpc** [comment]
I was also wondering if rolling checksum `struct`s should reset themselves when finding the edge.

**Comment by aidanhs** [comment]
> I was also wondering if rolling checksum structs should reset themselves when finding the edge.

I mentioned this in the first point - if you don't reset and you've not set a minimum chunk size, your content defined chunking become partially defined based on the contents of the previous chunk, which is almost certainly not what you want!

And yes, gear sounds pretty good (fastcdc possibly even better!) so migrating to that seems reasonable. I mainly wanted to raise this issue because of the iterator + minimum window size idea that sprung into my mind.

**Comment by dpc** [comment]
Yes, for a short window the state of the chunker is different, but:

* the previous chunk has a high chance of being the same as it was in previous runs
* the chunker state will get back to normal quickly, as such is a nature of rolling hashes
* the worst that can happen is finding a needless egde, generating a chunk, which has a very small overhead in the great scheme of things

My point was more about: it is actually not obvious should the chunker state be reset or not. The difference so so tiny. And if it really should, and my change in rdedup is kind of wrong, then maybe it should be done automatically by the chunker struct anyway.
